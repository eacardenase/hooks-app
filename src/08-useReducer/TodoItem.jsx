import PropTypes from 'prop-types';

export const TodoItem = ({ todo, handleRemoveTodo, handleToggleTodo }) => {
    return (
        <li className="list-group-item d-flex justify-content-between">
            <span
                className={`align-self-center ${
                    todo.done ? 'text-decoration-line-through fst-italic' : ''
                }`}
                onClick={() => handleToggleTodo(todo.id)}
                aria-label="todo-description"
            >
                {todo.description}
            </span>

            <button
                className="btn btn-danger"
                onClick={() => handleRemoveTodo(todo.id)}
            >
                Borrar
            </button>
        </li>
    );
};

TodoItem.propTypes = {
    todo: PropTypes.object.isRequired,
    handleRemoveTodo: PropTypes.func.isRequired,
    handleToggleTodo: PropTypes.func.isRequired,
};
