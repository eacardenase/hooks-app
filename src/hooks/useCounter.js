import { useState } from 'react';

export const useCounter = (initialValue = 10) => {
    const [counter, setCounter] = useState(initialValue);

    const incrementValue = (value = 1) =>
        setCounter((currCounter) => currCounter + value);
    const decreaseValue = (value = 1) => {
        if (counter - value < 0 || counter === 0) {
            return;
        }

        setCounter((currCounter) => currCounter - value);
    };
    const resetValue = () => setCounter(initialValue);

    return {
        counter,
        incrementValue,
        decreaseValue,
        resetValue,
    };
};
