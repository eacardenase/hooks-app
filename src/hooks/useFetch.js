import { useEffect, useState } from 'react';

export const useFetch = (url) => {
    const [state, setState] = useState({
        data: null,
        isLoading: true,
        hasError: false,
    });

    const getFetch = async () => {
        setState({
            ...state,
            isLoading: true,
        });

        try {
            const response = await fetch(url);
            const data = await response.json();

            setState({
                data,
                isLoading: false,
                hasError: false,
            });
        } catch (error) {
            setState({
                data: null,
                isLoading: false,
                hasError: true,
            });
        }
    };

    useEffect(() => {
        getFetch();
    }, [url]);

    return {
        data: state.data,
        isLoading: state.isLoading,
        hasError: state.hasError,
    };
};
