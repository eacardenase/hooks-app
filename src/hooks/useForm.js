import { useState } from 'react';

export const useForm = (initialForm = {}) => {
    const [formState, setFormState] = useState(initialForm);

    const handleInputChange = ({ target }) => {
        const { name, value } = target;

        setFormState({
            ...formState,
            [name]: value,
        });
    };

    const handleFormReset = () => {
        setFormState(initialForm);
    };

    return { ...formState, formState, handleInputChange, handleFormReset };
};
