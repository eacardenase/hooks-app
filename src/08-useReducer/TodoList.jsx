import PropTypes from 'prop-types';

import { TodoItem } from './TodoItem';

export const TodoList = ({ todos, handleRemoveTodo, handleToggleTodo }) => {
    return (
        <ul className="list-group">
            {todos.map((todo) => (
                <TodoItem
                    key={todo.id}
                    todo={todo}
                    handleRemoveTodo={handleRemoveTodo}
                    handleToggleTodo={handleToggleTodo}
                />
            ))}
        </ul>
    );
};

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleRemoveTodo: PropTypes.func.isRequired,
};
