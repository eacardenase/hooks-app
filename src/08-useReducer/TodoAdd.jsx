import PropTypes from 'prop-types';

import { useForm } from '../hooks/useForm';

export const TodoAdd = ({ handleNewTodo }) => {
    const { description, handleInputChange, handleFormReset } = useForm({
        description: '',
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.length < 2) {
            return;
        }

        const newTodo = { id: new Date().getTime(), description, done: false };

        handleNewTodo(newTodo);
        handleFormReset();
    };

    return (
        <form onSubmit={handleSubmit} aria-label="form">
            <input
                type="text"
                placeholder="Que debo hacer?"
                className="form-control"
                name="description"
                value={description}
                onChange={handleInputChange}
            />
            <button type="submit" className="btn btn-outline-primary mt-1">
                Agregar
            </button>
        </form>
    );
};

TodoAdd.propTypes = {
    handleNewTodo: PropTypes.func.isRequired,
};
