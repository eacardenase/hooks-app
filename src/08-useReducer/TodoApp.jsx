import { useCallback, useEffect, useReducer } from 'react';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';

import { todoReducer } from './todoReducer';

const initialState = [];

const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const TodoApp = () => {
    const [todos, dispatch] = useReducer(todoReducer, initialState, init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleNewTodo = useCallback((todo) => {
        const action = { type: '[TODO] Add Todo', payload: todo };

        dispatch(action);
    }, []);

    const handleRemoveTodo = useCallback((id) => {
        const action = {
            type: '[TODO] Remove Todo',
            payload: id,
        };

        dispatch(action);
    }, []);

    const handleToggleTodo = useCallback((id) => {
        const action = {
            type: '[TODO] Toggle Todo',
            payload: id,
        };

        dispatch(action);
    }, []);

    return (
        <>
            <h1>
                TodoApp: 10, <small>pendientes: 2</small>
            </h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleRemoveTodo={handleRemoveTodo}
                        handleToggleTodo={handleToggleTodo}
                    />
                </div>
                <div className="col-5">
                    <h4>Agregar TODO</h4>
                    <hr />

                    {/* onNewTodo() */}
                    <TodoAdd handleNewTodo={handleNewTodo} />
                </div>
            </div>
        </>
    );
};
