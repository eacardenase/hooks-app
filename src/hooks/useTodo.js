import { useCallback, useEffect, useReducer, useState } from 'react';

import { todoReducer } from '../08-useReducer/todoReducer';

const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const useTodo = ({ initialState = [] }) => {
    const [todos, dispatch] = useReducer(todoReducer, initialState, init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleNewTodo = useCallback((todo) => {
        const action = { type: '[TODO] Add Todo', payload: todo };

        dispatch(action);
    }, []);

    const handleRemoveTodo = useCallback((id) => {
        const action = {
            type: '[TODO] Remove Todo',
            payload: id,
        };

        dispatch(action);
    }, []);

    const handleToggleTodo = useCallback((id) => {
        const action = {
            type: '[TODO] Toggle Todo',
            payload: id,
        };

        dispatch(action);
    }, []);

    return {
        todos,
        todosCount: todos.length,
        pendingTodosCount: todos.filter((todo) => !todo.done).length,
        handleNewTodo,
        handleRemoveTodo,
        handleToggleTodo,
    };
};
