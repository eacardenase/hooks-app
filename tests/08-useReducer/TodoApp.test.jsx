import { describe, test, expect, vi, beforeEach } from 'vitest';

import { cleanup, render, screen } from '@testing-library/react';

import { TodoApp } from '../../src/08-useReducer/TodoApp';
import { useTodo } from '../../src/hooks/useTodo';

vi.mock('../../src/hooks/useTodo');

describe('<TodoApp />', () => {
    beforeEach(() => {
        cleanup();
    });

    useTodo.mockReturnValue({
        todos: [
            {
                id: 1,
                description: 'Todo #1',
                done: false,
            },
            {
                id: 2,
                description: 'Todo #2',
                done: true,
            },
        ],
        todosCount: 2,
        pendingTodosCount: 1,
        handleNewTodo: vi.fn(),
        handleRemoveTodo: vi.fn(),
        handleToggleTodo: vi.fn(),
    });

    test('should match snapshot', () => {
        const { container } = render(<TodoApp />);

        // screen.debug();

        expect(container).toMatchSnapshot();
    });

    test('should show component correctly', () => {
        render(<TodoApp />);

        expect(screen.getByText('Todo #1')).toBeDefined();
        expect(screen.getByText('Todo #2')).toBeDefined();
        expect(screen.getByRole('textbox')).toBeDefined();
    });
});
