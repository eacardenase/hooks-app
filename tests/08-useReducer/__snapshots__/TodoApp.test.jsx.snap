// Vitest Snapshot v1

exports[`<TodoApp /> > should match snapshot 1`] = `
<div>
  <h1>
    TodoApp: 
    2
    ,
    <small>
      pendientes:
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="todo-description"
            class="align-self-center "
          >
            Todo #1
          </span>
          <button
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="todo-description"
            class="align-self-center text-decoration-line-through fst-italic"
          >
            Todo #2
          </span>
          <button
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          class="form-control"
          defaultvalue=""
          name="description"
          placeholder="Que debo hacer?"
          type="text"
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
